packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -2,35 +2,139 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-
 import 'dart:async';
 import 'dart:ui' as ui show Image, Codec, FrameInfo;
 import 'dart:ui' show hashValues;
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/scheduler.dart';
 
+/// A reference to a [ui.Image] object.
+///
+/// To get an ImageHandle, clients should call [ImageInfo.obtainImageHandle].
+///
+/// Clients of an [ImageStream] should call [dispose] on this object when they
+/// will no longer need to use the [ImageInfo.image] vended to them in
+/// [ImageStreamCompleter.setImage]. When all clients of the stream have
+/// disposed their handles, the underlying [ui.Image] will be disposed. Clients
+/// must not call dispose more than once on the reference they receive from
+/// [ImageInfo.obtainImageHandle].
+///
+/// Clients should _not_ call [ui.Image.dispose] directly, as it will prevent
+/// other clients from using the same image reference should they need to.
+class ImageHandle {
+  /// Do not use this, use [_create] instead to make sure that ImageHandles
+  /// for the same [ui.Image] are the same.
+  ImageHandle._(this._image)
+    : assert(_image != null);
+
+  factory ImageHandle._create(ui.Image image) {
+    assert(image != null);
+    return _handles[image] ??= ImageHandle._(image);
+  }
+
+  static final Expando<ImageHandle> _handles = Expando<ImageHandle>();
```

Comment:
Why an Expando? Instead of its magic, shouldn't we always know when something needs to get dropped out of it? (e.g. when the handle ref count gets to zero?)

Reply:
Because we don't want to keep the `ImageHandle` or `Image`s this stores alive.

The problem here is that if multiple `ImageStream`s are vending the same `Image`, e.g. because two different image providers are resolving to the same image. We want to make sure that a single `Image` always uses the same `ImageHandle`.